library/src/main/java/com/voysis/sevice/local/loader/AssetsLoader.kt

```
@@ -0,0 +1,35 @@
+package com.voysis.sevice.local.loader
+
+import android.content.Context
+import java.io.File
+import java.io.FileOutputStream
+import java.io.IOException
+
+interface AssetsLoader {
+    fun extractAssets(fName: String): String
+}
+
+class LocalAssetsLoader(var context: Context) : AssetsLoader {
+
+    override fun extractAssets(fName: String): String {
+        val newLoc = context.getFilesDir().toString().plus(fName)
+        val f = File(newLoc)
+        if (!f.exists()) {
+            try {
+                val inputstream = context.getAssets().open(fName)
+                val fileOutputStream = FileOutputStream(f)
+
+                val buf = ByteArray(1024)
+                var len: Int = inputstream.read(buf)
+                while (len > 0) {
+                    fileOutputStream.write(buf, 0, len)
+                    len = inputstream.read(buf)
+                }
+                fileOutputStream.close()
+            } catch (e: IOException) {
+                e.printStackTrace()
+            }
```

we should probably throw a specific voysis exception here to indicate that there was a problem with the models the user included in their application.

---

I think this is correct here to be catching the IOException and printing the trace, it is just IO that's happening after all.

---

I suppose. catching a model exception should probably happen at the point where we try to load the models if that method can throw an error. Wouldn't use `e.printStackTrace` though. it prints to the device logcat without including the calling application. Better to use the android logger. https://stackoverflow.com/a/2073844/2342568

---

It's not correct to swallow the exception here. You catch an exception if you can do something about it. Here, nothing can be done - the asset extraction failed so the rest of the application cannot run (as it will not be able to load the models.) If you cannot do anything about it, don't catch the exception, let it bubble out so that higher levels in the call stack can deal with it.